### PR TITLE
enhance test support and update realtime test to use serverTime

### PR DESCRIPTION
**Pull request by rasahner**
Currently we are running the community integration tests during our builds, along with some tests of our own.  The test code we are using is a copy of what is in druid-io/druid/integration-tests with some local modifications and additions.  We would like, instead, to base our test code on jars made from what's in druid-io/druid/integration-tests.  To do so, we need some generalizations of the test support code.  This PR proposes those, and it also proposes an enhancement to one of the tests.

We do not run the tests using mvn.  That's because some of the test systems we use do not have access to the external internet, and in cases where they do have access, doing downloads can be prohibitively slow.  Instead, we run the tests by setting an appropriate classpath and executing "java org.testng.TestNG".  For this to work, we need to have the compiled code from both integration-tests/src/main and integration-tests/src/test on the classpath.  The code from main is available in druid-integration-tests-n.n.n.jar.  We'd like to enhance pom.xml to get a jar with the code from integration-tests/src/main.

We have a receiver firehose test that uses rejection policy serverTime and creates data events on the fly with current datestamps.  It sends events that are out of order (but within the window) and events that are outside (before) the window.  It does queries using the router.  I folded our test into ITRealtimeIndexTaskTest.

The test wants to build queries and expected results based on the timestamp, and drive the query test from a String instead of from the contents of a file.  I moved FromFileTestQueryHelper to TestQueryHelper, and gave it two functions for doing the query, one taking the String with the query information, the other taking the name of a file from which to read the String.  I also wanted to allow tests to query either the broker or the router, so I added a url parameter to the query functions.  (I left the broker as the default, for backward compatibility.)

The new version of ITRealtimeIndexTaskTest can't share resources with ITUnionQueryTest any more; I separated the resources.

Because we're sending data with current timestamps, we want to be able to unload and kill segments with timestamps other than in 2013.  I added an unloadAndKillData function that takes start and end timestamps as parameters.

We were seeing some tests fail because they were doing queries before all of the data was available.  Waiting until "loadstatus?simple" said 0 for the datasource of interest was not sufficient.  Experiments with the twitter query test showed that sometimes when loadstatus?simple showed "twitterstream":0, intervals?full showed only two of the three segments. At that point the test's queries were failing in a way that was consistent with two of the three segments being present.  After the intervals query showed all the segments present, the queries started passing.  If we wait until plain "loadstatus" says 100.0, all of the segments should be present.

The areSegmentsLoaded function is also used to determine when segments are not available any more after they are unloaded.  The change to areSegmentsLoaded() will not change the ability to use it for that purpose.

We are starting to have some tests that use additional configuration information.  I want to add a function getHistoricalHost() to the configuration classes, and a generic function getProperty() for other properties. (This PR changes the configuration file string for the middle manager because it was incorrect before.)

When we run the wikipedia query tests, some of the queries often time out.  I increased the timeout value for some of the queries in wikipedia_editstream_queries.jon.

We have an integration test that uses the kafka firehose, and one that uses tranquility.  We hope to contribute those also, once some details are taken care of.

### CODE REVIEW

File: integration-tests/src/main/java/io/druid/testing/DockerConfigProvider.java [hunk 1]
```diff
@@ -69,6 +75,12 @@ public String getZookeeperHosts()
       {
         return dockerIp + ":2181";
```
Comment (fjy) [inline integration-tests/src/main/java/io/druid/testing/DockerConfigProvider.java]: same here

File: integration-tests/src/main/java/io/druid/testing/DockerConfigProvider.java [hunk 2]
```diff
@@ -59,6 +59,12 @@ public String getBrokerHost()
       }
 
       @Override
+      public String getHistoricalHost()
+      {
+        return dockerIp + ":8083";
```
Comment (fjy) [inline integration-tests/src/main/java/io/druid/testing/DockerConfigProvider.java]: it might be nice to make these configurable

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java [hunk 3]
```diff
@@ -68,9 +70,14 @@ private String getBrokerURL()
 
   public List<Map<String, Object>> query(Query query)
   {
+      return query (getBrokerURL(), query);
```
Comment (fjy) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: formatting (no space)

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java [hunk 4]
```diff
@@ -40,6 +41,7 @@
 
 public class QueryResourceTestClient
 {
+  private final static Logger LOG = new Logger(QueryResourceTestClient.class);
```
Comment (fjy) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: can we lowercase to match styling elsewhere in druid?

File: integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
+/*
+ * Druid - a distributed column store.
```
Comment (fjy) [inline integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java]: we should use the other license header for companies that are not mmx

File: integration-tests/src/test/resources/queries/wikipedia_editstream_queries.json [hunk 6]
```diff
@@ -373,7 +373,7 @@
             "context": {
                 "useCache": "true",
                 "populateCache": "true",
-                "timeout": 60000
+                "timeout": 180000
```
Comment (fjy) [inline integration-tests/src/test/resources/queries/wikipedia_editstream_queries.json]: this is a long timeout, any particular reason why it is needed?

File: integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java [hunk 7]
```diff
@@ -132,10 +205,56 @@ public void postEvents() throws Exception
...
+          );
+          int eventsPosted = client.postEvents(events, this.jsonMapper, MediaType.APPLICATION_JSON);
+          if (eventsPosted != events.size()) {
+              throw new ISE("Event not posted");
+          }
+          
+          try {
+              Thread.sleep(DELAY_BETWEEN_EVENTS_SECS * 1000);
+          } catch (InterruptedException ex) { /* nothing */ }
+          dtLast = dt;
```
Comment (fjy) [inline integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java]: There's a bunch of formatting issues. Make sure to use the IntelliJ style guide

File: integration-tests/src/main/java/io/druid/testing/DockerConfigProvider.java (hunk 2 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/DockerConfigProvider.java]: agreed - that's why a previous PR added ConfigFileConfigProvider, which allows host and port to be configured for each node type, and other general configuration information.

File: integration-tests/src/main/java/io/druid/testing/DockerConfigProvider.java (hunk 1 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/DockerConfigProvider.java]: same - ConfigFileConfigProvider

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java (hunk 3 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: removed space

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java (hunk 4 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: lowercased

File: integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java (hunk 5 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java]: replace with a different license header - hope it's the one you meant

File: integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java (hunk 7 above)
Comment (rasahner) [inline integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java]: reformatted whole file using IntelliJ

File: integration-tests/src/test/resources/queries/wikipedia_editstream_queries.json (hunk 6 above)
Comment (rasahner) [inline integration-tests/src/test/resources/queries/wikipedia_editstream_queries.json]: with shorter timeouts, this part of the test was failing.  I asked about this a while back at the dev meeting and the answer seemed to be that we're not trying to measure performance with these tests and to not be concerned about it unless it was outrageous.  I made the timeout much bigger so that we'd quit seeing what would be considered spurious failures.  If we think there might be a problem with this topN query, perhaps we could investigate it under a separate issue.

Comment (drcrallen) [inline integration-tests/src/test/resources/queries/wikipedia_editstream_queries.json]: I remember that conversation and still stand by that assessment. The clarification here for posterity is appreciated :)

**Comment by fjy** [comment]
:+1:

File: integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  {
+    LOG.info("Starting failure query tests for [%s]", filePath);
+    List<QueryWithResults> queries =
+        jsonMapper.readValue(
+            TestQueryHelper.class.getResourceAsStream(filePath),
+            new TypeReference<List<QueryWithResults>>()
+            {
+            }
+        );
+    testFailureQueries (queries, timesToRun);
```
Comment (nishantmonu51) [inline integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java]: formatting.

Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java]: reformatted TestQueryHelper with IntelliJ

**Comment by rasahner** [comment]
Are there any other things I should address?

File: integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java [hunk 9]
```diff
@@ -61,10 +64,18 @@ private void loadProperties(String configFile)
     }
     routerHost = props.get("router_host") + ":" + props.get("router_port");
     brokerHost = props.get("broker_host") + ":" + props.get("broker_port");
+    historicalHost = props.get("historical_host") + ":" + props.get("historical_port");
     coordinatorHost = props.get("coordinator_host") + ":" + props.get("coordinator_port");
     indexerHost = props.get("indexer_host") + ":" + props.get("indexer_port");
-    middleManagerHost = props.get("middle_manager_host");
+    middleManagerHost = props.get("middlemanager_host");
```
Comment (himanshug) [inline integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java]: will this change not break existing properties file? why is this needed?

File: integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java [hunk 10]
```diff
@@ -61,10 +64,18 @@ private void loadProperties(String configFile)
...
     zookeeperHosts = props.get("zookeeper_hosts");
+
+    LOG.info ("router: " + routerHost);
+    LOG.info ("broker: " + brokerHost);
+    LOG.info ("coordinator: " + coordinatorHost);
+    LOG.info ("overlord: " + indexerHost);
+    LOG.info ("middle manager: " + middleManagerHost);
+    LOG.info ("zookeepers: " + zookeeperHosts);
```
Comment (himanshug) [inline integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java]: can you do LOG.info("xxx: [%s]", xxx) in all these?

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java [hunk 11]
```diff
@@ -93,6 +100,7 @@ private String getBrokerURL()
       );
     }
     catch (Exception e) {
+      log.error("exception: " + e.getMessage());
```
Comment (himanshug) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: if the exception is propagated then why is the log needed, whoever handles the exception will log ?however, if we do log, then we should probably do `log.error(e, "exception")`

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java [hunk 12]
```diff
@@ -40,6 +41,7 @@
 
 public class QueryResourceTestClient
 {
+  private final static Logger log = new Logger(QueryResourceTestClient.class);
```
Comment (himanshug) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: in pretty much all other places in this PR, it is called uppercase LOG . I would keep it consistent and @fjy I don't actually mind it being uppercase given that this is a static final .

File: integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+      LOG.info("Running Query %s", queryWithResult.getQuery().getType());
+      try {
+        List<Map<String, Object>> result = queryClient.query(queryWithResult.getQuery());
+        throw new ISE(
+            "No error while querying[%s]",
+            queryWithResult.getQuery()
+        );
+      }
+      catch (ISE e) {
+        LOG.info("query failed, as expected");
```
Comment (himanshug) [inline integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java]: wouldn't you come here even if query was successful and ISE is really from the line 134?

File: integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java [hunk 14]
```diff
@@ -99,15 +168,17 @@ public Boolean call() throws Exception
           "Real-time generated segments loaded"
       );
 
-      // run queries on historical nodes
-      this.queryHelper.testQueriesFromFile(INDEX_QUERIES_RESOURCE, 2);
+      // queries should be answered by historical
+      this.queryHelper.testQueriesFromString(getRouterURL(), queryStr, 2);
     }
     catch (Exception e) {
       e.printStackTrace();
```
Comment (himanshug) [inline integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java]: can you remove this e.printStackTrace()?

File: integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java [hunk 15]
```diff
@@ -132,10 +213,57 @@ public void postEvents() throws Exception
...
+      dtLast = dt;                       // timestamp of last event
+      String line;
+      while ((line = reader.readLine()) != null) {
+        if (i == 15) { // for the 15th line, use a time before the window
+          dt = dt.minusMinutes(10);
+        } else if (i == 18) { // use a time 6 seconds ago so it will be out of order
+          dt = dt.minusSeconds(6);
+        }
+        String event = line.replace(TIME_PLACEHOLDER, EVENT_FMT.print(dt));
+        LOG.info("sending event: \n" + event);
```
Comment (himanshug) [inline integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java]: can you use formatter instead of string concatenation? (here and in all other places if applicable)

File: integration-tests/src/test/java/io/druid/tests/indexer/ITUnionQueryTest.java [hunk 16]
```diff
@@ -42,7 +42,7 @@
 public class ITUnionQueryTest extends AbstractIndexerTest
 {
   private static final Logger LOG = new Logger(ITUnionQueryTest.class);
-  private static final String REALTIME_TASK_RESOURCE = "/indexer/wikipedia_realtime_index_task.json";
+  private static final String REALTIME_TASK_RESOURCE = "/indexer/wikipedia_union_index_task.json";
```
Comment (himanshug) [inline integration-tests/src/test/java/io/druid/tests/indexer/ITUnionQueryTest.java]: can you change the variable name also to UNION_TASK_RESOURCE?

**Comment by nishantmonu51** [comment]
+1, Also verified that the integration-tests pass with the above changes.

File: integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java (hunk 10 above)
Comment (drcrallen) [inline integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java]: +1

File: integration-tests/src/main/java/io/druid/testing/clients/CoordinatorResourceTestClient.java [hunk 17]
```diff
@@ -70,7 +70,7 @@ private String getCoordinatorURL()
   {
     Map<String, Integer> status = null;
     try {
-      StatusResponseHolder response = makeRequest(HttpMethod.GET, getCoordinatorURL() + "loadstatus?simple");
+      StatusResponseHolder response = makeRequest(HttpMethod.GET, getCoordinatorURL() + "loadstatus");
```
Comment (drcrallen) [inline integration-tests/src/main/java/io/druid/testing/clients/CoordinatorResourceTestClient.java]: Since this is being touched, can you make it use the URL building methods instead of string building?

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java (hunk 11 above)
Comment (drcrallen) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: ^^^^^ +1

File: integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java (hunk 9 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java]: This was incorrect before, it should have been middlemanager_host all along to be consistent with the properties file we are using at Y!.  I'd be surprised if anybody is using this configuration method yet, but if you think somebody might be, I can change the code to look for either string - or I can leave it as is and change our Y! code later.

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java (hunk 11 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: Without the log message, the information about the failure appears only in the testng test results information, not in the test output.  I find it extremely convenient when diagnosing problems to have information about the problem in both places.  But since the log messages has\ve caused repeated code review comments, I will cease doing that.

Comment (drcrallen) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: That is a reasonable excuse to include a log message here, but it should still conform to `log.error(e, "message")`

File: integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java (hunk 9 above)
Comment (himanshug) [inline integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java]: if it has always been wrong then, it is fine. This change gave me the impression that we accidendally changed a property name.

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java (hunk 11 above)
Comment (himanshug) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: In my experience, an exception usually passes through a big stack of method calls and we put the log inside every method then pretty much same log(and trace) gets printed in the log files many many times and creates unnecessary noise. That is why, I usually do not log it if it is being propagated.
We might have a special case here but not too sure. I would not block the PR on this but this kind of logging should be exception than the norm.

Comment (drcrallen) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: @himanshug I agree with that assessment, but I also agree with @rasahner that merging logs when digging into errors is annoying. Would it appease you to have this log at a debug level?

Comment (himanshug) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: we can keep it here for this special reason at error level itself, but in general wouldn't log exceptions when propagating them.
pls change to `log.error(e, "message")` however.

File: integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java (hunk 10 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/ConfigFileConfigProvider.java]: done

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java (hunk 11 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: removed log message

File: integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java (hunk 12 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/clients/QueryResourceTestClient.java]: variable not used any more, removed

File: integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java (hunk 13 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/utils/TestQueryHelper.java]: yeah, this is broken.  it's not used in integration-tests - I've removed it for now, will bring back a fixed one when it becomes needed.

File: integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java (hunk 14 above)
Comment (rasahner) [inline integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java]: removed

File: integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java (hunk 15 above)
Comment (rasahner) [inline integration-tests/src/test/java/io/druid/tests/indexer/ITRealtimeIndexTaskTest.java]: modified

File: integration-tests/src/test/java/io/druid/tests/indexer/ITUnionQueryTest.java (hunk 16 above)
Comment (rasahner) [inline integration-tests/src/test/java/io/druid/tests/indexer/ITUnionQueryTest.java]: changed

File: integration-tests/src/main/java/io/druid/testing/clients/CoordinatorResourceTestClient.java (hunk 17 above)
Comment (rasahner) [inline integration-tests/src/main/java/io/druid/testing/clients/CoordinatorResourceTestClient.java]: added and used URL building method

**Comment by himanshug** [comment]
LGTM (once travis is done)

**Comment by himanshug** [comment]
@drcrallen this is +1 from everyone who reviewed this PR , pls merge if your comments are addressed. thanks.
